---
{title: "Minado de datos biologicos", subtitle: "RNAseq (parte II) y Single-Cell RNA" ,
date: '2020-11-19',
author: "Alejandro Acevedo-Aracena, _PhD_^[_FONDECYT Postdoctoral Fellow_, Universidad de Chile, deepen.data@gmail.com]; Manuel Muñoz-González^[_Pregrado_, Universidad de Chile]", output: {html_document: {theme: flatly, highlight: tango, toc: true, toc_float: {collapsed: false, smooth_scroll: true}, df_print: paged, code_folding: show}, html_notebook: default}}
---

### Case: RNAseq read count data
```{r message=FALSE, warning=FALSE, eval=F}
install.packages("remotes", dependencies = T)
remotes::install_github("WTaoUMC/RegEnrich", dependencies = T)
```
#  Background
Here we show how to apply RegEnrich on the RNAseq data by analyzing Kang et al’s monocyte-macrophage-IFN stimulation dataset ( GSE130567). There are multiple experiment conditions in this study. But here we would like to focus on partial samples in which monocytes were cultured with 10 ng/ml human macrophage colonystimulating factor (M-CSF) in the presence (IFN-γ-primed macrophages) or absence (resting macrophages) of 100 U/ml human IFN-γ for 48 h. RNA were extracted and reverse transcripted followed by sequencing (50 bp, paired-end) using Illumina HiSeq 4000. Sequenced reads were mapped to reference human genome (hg19 assembly) using STAR aligner with default parameters. We will use the raw HT-seq counts for the RegEnrich analysis.
```{r}
library(RegEnrich)
library(GEOquery)
library(tidyverse)
eset <- getGEO(GEO = "GSE130567")[[1]] #ExpressionSet unpacked
pdata <- pData(eset)[, c("title", "geo_accession", "cultured in:ch1", "treatment:ch1")]
colnames(pdata) <- c("title", "accession", "cultured", "treatment")
pData(eset) <- pdata
```

```{r}
# Estructura: eset[featureData (transcritos o genes) , phenoData (muestras: GSM)]
# Only samples cultured with M-CSF in the presence or absence of IFN-γ

eset <- eset[, pdata$treatment %in% c("NT", "IFNG-3h") & pdata$cultured == "M-CSF"]
```

```{r}
# Sample information (muestras: GSM)
sampleInfo           <- pData(eset)
rownames(sampleInfo) <- paste0(rep(c("Resting", "IFNG"), each = 3), 1:3)
sampleInfo$treatment <- factor(rep(c("Resting", "IFNG"), each = 3),
                              levels = c("Resting", "IFNG"))
```
Download read count file and decompose into a temporary folder.
Desde https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE130567 Bajamos GSE130567_RAW.tar
Botón derecho - copiar ubicación del enlace:
https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE130567&format=file

```{r}
tmpFolder = tempdir()
tmpFile   = tempfile(pattern = "GSE130567_", tmpdir = tmpFolder, fileext = ".tar")
download.file("https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE130567&format=file",
              destfile = tmpFile, mode = "wb")

untar(tmpFile, exdir = tmpFolder)
files = untar(tmpFile, list = TRUE)
filesFull = file.path(tmpFolder, files)
```
Then read the raw read counts in these files.

(usaremos funciones de R base preferetement)
```{r}
dat <- list()
for (file in filesFull){
  accID<-gsub(".*(GSM\\d+).*", "\\1", file) # reprint using \\1 
  if(accID %in% sampleInfo$accession){
    zz<-gzfile(file, "rt") #A connection with  
    zzdata<-read.csv(zz, header = FALSE, sep = "\t", skip = 4, row.names = 1)
    close(zz)
    zzdata<-zzdata[,1, drop = FALSE] # Extract the first numeric column
    colnames(zzdata)<-accID
    dat<-c(dat, list(zzdata)) #Grow a list
  }
}

# lo mismo que purrr::reduce(dat,cbind)

edata0 = do.call(cbind, dat)

edata = edata0[grep(".*[0-9]+$", rownames(edata0)),] # remove PAR locus genes
#check
rownames(edata0)[!rownames(edata0) %in% rownames(edata)]

```
```{r}
rownames(edata) <- substr(rownames(edata), 1, 15)
colnames(edata) <- rownames(sampleInfo)
```
```{r}
# Retain genes with average read counts higher than 1
edata <- edata[rowMeans(edata) > 1,]
```
ere we randomly take only 5,000 genes to quickly illustrate how to use RegEnrich, but to see the real result from the analysis, you should neglect the following step.

```{r}
set.seed(1234)
edata <- edata[sample(1:nrow(edata), 5000), ]
expressionMatrix = as.matrix(edata) 
expressionMatrix %>% dim
```
# Get regulators
Database of transcription co-factors and transcription factor interactions:
https://tools.sschmeier.com/tcof/home/
http://www.yeastract.com/consensuslist.php
https://www.mrc-lmb.cam.ac.uk/genomes/FlyTF/old_index.html

```{r}
data(TFs)
sample(TFs)
unique(TFs$TF) -> human_regulators
```

# Initializing a RegenrichSet object
```{r}
RegenrichSet_Base = RegenrichSet(expr = expressionMatrix, # expression data (matrix)
                      colData = sampleInfo, # sample information (data frame)
                      reg = human_regulators, # regulators
                      method = "LRT_DESeq2", # differential expression analysis method
                      design = ~ treatment, # desing fomula
                      reduced = ~ 1, # reduced
                      networkConstruction = "COEN", # network inference method
                      enrichTest = "FET") # enrichment analysis method

```
# Differential expression analysis
```{r}
RegenrichSet_Base          %>% regenrich_diffExpr -> object_diffExpr
object_diffExpr %>% results_DEA
```
# Regulator-target network inference
```{r}
object_diffExpr %>% regenrich_network  -> object_network
object_network %>% results_topNet
```
```{r}
slot(RegenrichSet_Base, "paramsIn")$reg -> all_regulators
all_regulators %>% str

object_network@topNetwork@set$set -> regulators_found
regulators_found %>% str

regulators_found %in% all_regulators


```

### GRN (based on random forest)
```{r}
library(BiocParallel)
# on non-Windows computers (use 2 workers)
bpparam = register(MulticoreParam(workers = 8, RNGseed = 1234))
# on Windows computers (use 2 workers)
# bpparam = register(SnowParam(workers = 2, RNGseed = 1234))
# the lower minR is set, the less edges and potential less regulators are retained.
object_GRN_multicore_network = regenrich_network(object_diffExpr, networkConstruction = "GRN", 
                              BPPARAM = bpparam, minR = 0.3)
```
```{r}
object_GRN_multicore_network %>% results_topNet
```
# Enrichment analysis


```{r}
object_network  %>% regenrich_enrich   -> object_enrich
object_enrich %>%results_enrich  
slot(results_enrich(object_enrich), "allResult") -> enrich_FET

enrich_FET$ID %in% all_regulators
enrich_FET
```


# Gene set enrichment analysis (GSEA)
```{r}
set.seed(123)
object_enrich_GSEA <- regenrich_enrich(object_network, enrichTest = "GSEA", nperm = 5000)
enrich_GSEA <- slot(results_enrich(object_enrich_GSEA), "allResult")
enrich_GSEA
```
```{r}
plotOrders(enrich_FET[[1]], enrich_GSEA[[1]])
```
# Regulator scoring and ranking
The RegEnrich score is a summarized information from both differential expression analysis and regulator enrichment analysis for regulators. This step of RegEnrich analysis is done by regenrich_rankScore function.

Above all, the differential expression analysis is performed by Limma method, regulator-target network is infered by COEN method, and enrichment analysis is performed by FET method, so the scores and ranking summarize the importance of regulators by considering regulatory interactions in the studied biological process.
```{r}

object_enrich   %>% regenrich_rankScore -> object_rankScore
res_score = results_score(object_rankScore)

results_score(object_rankScore)
```


```{r}
object_rankScore_GSEA <- regenrich_rankScore(object_enrich_GSEA)
results_score(object_rankScore_GSEA)
```

```{r}
plotRegTarExpr(object_rankScore, reg = "ENSG00000143437")
plotRegTarExpr(object_rankScore_GSEA, reg = "ENSG00000239306")
```

### More mining

```{r}
enrich_FET
```
```{r message=FALSE, warning=FALSE}
library(KEGGREST)
library(tidyverse)
KEGGREST::listDatabases() -> all.kegg.databases
```

```{r}
library(org.Hs.eg.db)
org.Hs.eg.db %>% keytypes

enrich_FET.ENTREZID             <- clusterProfiler::bitr(enrich_FET$ID, "ENSEMBL", "ENTREZID", "org.Hs.eg.db")
enrich_FET.SYMBOL             <- clusterProfiler::bitr(enrich_FET$ID, "ENSEMBL", "SYMBOL", "org.Hs.eg.db")

enrich_FET.ENTREZID.SYMBOL <-  inner_join(enrich_FET.ENTREZID, enrich_FET.SYMBOL, by = 'ENSEMBL')
```

```{r}

Regulators_ego <- clusterProfiler::enrichGO(gene         = enrich_FET.ENTREZID.SYMBOL$ENTREZID,
                OrgDb         = org.Hs.eg.db,
                keyType       = 'ENTREZID',
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.01,
                qvalueCutoff  = 0.01)
Regulators_ego@result    
```

```{r}
inner_join(enrich_FET.ENTREZID.SYMBOL, enrich_FET , by = c('ENSEMBL'='ID')) -> enrich_FET_full
enrich_FET_full %>% .[1,] %>% .[['SYMBOL']] 
enrich_FET_full %>% .[1,] %>% .[['geneID']]%>% str_split('/') %>% unlist() -> ensembl_ARNT

regulator_ARNT.ENTREZID             <- clusterProfiler::bitr(ensembl_ARNT, "ENSEMBL", "ENTREZID", "org.Hs.eg.db")

regulator_ARNT.SYMBOL             <- clusterProfiler::bitr(ensembl_ARNT, "ENSEMBL", "SYMBOL", "org.Hs.eg.db")

regulator_ARNT.ENTREZID.SYMBOL <-  inner_join(regulator_ARNT.ENTREZID, regulator_ARNT.SYMBOL, by = 'ENSEMBL')
```
```{r}

regulator_ARNT_ego <- clusterProfiler::enrichGO(gene         = regulator_ARNT.ENTREZID.SYMBOL$ENTREZID,
                OrgDb         = org.Hs.eg.db,
                keyType       = 'ENTREZID',
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.01,
                qvalueCutoff  = 0.01)
regulator_ARNT_ego@result    
```

# With RNA universe
```{r}

expressionMatrix %>% rownames -> all_rna_ensmble

all_rna_ensmble.ENTREZID             <- clusterProfiler::bitr(all_rna_ensmble, "ENSEMBL", "ENTREZID", "org.Hs.eg.db")

all_rna_ensmble.SYMBOL             <- clusterProfiler::bitr(all_rna_ensmble, "ENSEMBL", "SYMBOL", "org.Hs.eg.db")

all_rna_ensmble.ENTREZID.SYMBOL <-  inner_join(all_rna_ensmble.ENTREZID, all_rna_ensmble.SYMBOL, by = 'ENSEMBL')
```

```{r}

regulator_ARNT_vs_universe <- clusterProfiler::enrichGO(gene          = regulator_ARNT.ENTREZID.SYMBOL$ENTREZID,
                universe      = all_rna_ensmble.ENTREZID.SYMBOL$ENTREZID,
                OrgDb         = org.Hs.eg.db,
                keyType       = "ENTREZID",
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.01,
                qvalueCutoff  = 0.01,
        readable      = TRUE)

regulator_ARNT_vs_universe@result 
```

# Kegg modules
```{r}

regulator_ARNT_mkk <- clusterProfiler::enrichMKEGG(gene = regulator_ARNT.ENTREZID.SYMBOL$ENTREZID,
                    organism = "hsa",
                   pvalueCutoff = 0.01,
                   qvalueCutoff = 0.01)

regulator_ARNT_mkk@result
```
### ReactomePA

```{r}
#BiocManager::install("ReactomePA", checkBuilt = T)
reactome  <- ReactomePA::enrichPathway( regulator_ARNT.ENTREZID.SYMBOL$ENTREZID,
                          minGSSize = 10,  organism = "human", pvalueCutoff = 0.01)

reactome@result %>% dplyr::select(matches('descri'))
```


```{r}
all_reactome  <- ReactomePA::enrichPathway(enrich_FET.ENTREZID.SYMBOL$ENTREZID,
                          minGSSize = 5,  organism = "human", pvalueCutoff = 0.05)
all_reactome@result%>% dplyr::select(matches('descri'))
```
### Disease over-representation analysis
```{r}

x <- DOSE::enrichDO(gene          = enrich_FET.ENTREZID.SYMBOL$ENTREZID,
              ont           = "DO",
              pvalueCutoff  = 0.05,
              pAdjustMethod = "BH",
              minGSSize     = 5,
              maxGSSize     = 500,
              qvalueCutoff  = 0.05,
              readable      = FALSE)

x@result
```
### Over-representation analysis for the disease gene network


```{r}

dgn <-DOSE::enrichDGN(
  gene   = enrich_FET.ENTREZID.SYMBOL$ENTREZID,
  pvalueCutoff = 0.01,
  pAdjustMethod = "BH",
  qvalueCutoff = 0.01)


dgn %>% as.data.frame
```
### Biological theme comparison

Comparing multiple gene lists
```{r}
#regulator_ARNT.ENTREZID.SYMBOL
enrich_FET_full %>% .[2,] %>% .[['SYMBOL']] 
enrich_FET_full %>% .[2,] %>% .[['geneID']]%>% str_split('/') %>% unlist() -> ensembl_RBM14

ensembl_RBM14.ENTREZID             <- clusterProfiler::bitr(ensembl_RBM14, "ENSEMBL", "ENTREZID", "org.Hs.eg.db")
ensembl_RBM14.SYMBOL             <- clusterProfiler::bitr(ensembl_RBM14, "ENSEMBL", "SYMBOL", "org.Hs.eg.db")
regulator_RBM14.ENTREZID.SYMBOL <-  inner_join(ensembl_RBM14.ENTREZID, ensembl_RBM14.SYMBOL, by = 'ENSEMBL')
```
```{r}
library(tidyverse)
enrich_FET_full %>% .[3,] %>% .[['SYMBOL']] 
enrich_FET_full %>% .[3,] %>% .[['geneID']]%>% str_split('/') %>% unlist() -> ensembl_EHMT1

ensembl_EHMT1.ENTREZID             <- clusterProfiler::bitr(ensembl_EHMT1, "ENSEMBL", "ENTREZID", "org.Hs.eg.db")
ensembl_EHMT1.SYMBOL             <- clusterProfiler::bitr(ensembl_EHMT1, "ENSEMBL", "SYMBOL", "org.Hs.eg.db")
regulator_EHMT1.ENTREZID.SYMBOL <-  inner_join(ensembl_EHMT1.ENTREZID, ensembl_EHMT1.SYMBOL, by = 'ENSEMBL')
```
```{r}
library(tidyverse)
enrich_FET_full %>% .[4,] %>% .[['SYMBOL']] 
enrich_FET_full %>% .[4,] %>% .[['geneID']]%>% str_split('/') %>% unlist() -> ensembl_BIN1

ensembl_BIN1.ENTREZID             <- clusterProfiler::bitr(ensembl_BIN1, "ENSEMBL", "ENTREZID", "org.Hs.eg.db")
ensembl_BIN1.SYMBOL             <- clusterProfiler::bitr(ensembl_BIN1, "ENSEMBL", "SYMBOL", "org.Hs.eg.db")
regulator_BIN1.ENTREZID.SYMBOL <-  inner_join(ensembl_BIN1.ENTREZID, ensembl_BIN1.SYMBOL, by = 'ENSEMBL')
```

```{r}


my_regulated_genes.ENTREZID <- list(RBM14 = regulator_RBM14.ENTREZID.SYMBOL[['ENTREZID']], 
                                    ARNT  = regulator_ARNT.ENTREZID.SYMBOL[['ENTREZID']], 
                                    EHMT1 = regulator_EHMT1.ENTREZID.SYMBOL[['ENTREZID']],
                                    BIN1 = regulator_BIN1.ENTREZID.SYMBOL[['ENTREZID']])
```

```{r fig.height=5, fig.width=15}
library(clusterProfiler)
library(ReactomePA)

ck <- compareCluster(geneCluster = my_regulated_genes.ENTREZID, fun = 'enrichPathway',  pvalueCutoff=5)
dotplot(ck)
```


```{r fig.height=5, fig.width=15}
library(org.Hs.eg.db)
ck2 <- compareCluster(geneCluster = my_regulated_genes.ENTREZID, fun = enrichGO, OrgDb = org.Hs.eg.db, pvalueCutoff=5)
ck2 <- setReadable(ck2, OrgDb = org.Hs.eg.db, keyType="ENTREZID")
dotplot(ck2)
```

```{r fig.height=5, fig.width=15}
ck3 <- compareCluster(my_regulated_genes.ENTREZID, fun="enrichMKEGG",
                     organism="hsa", pvalueCutoff=5)
dotplot(ck3)
```

####################################################################
#scRNAseq

```{r}
#BiocManager::install("scRNAseq", checkBuilt = T)
#?ReprocessedFluidigmData
library(tidyverse)
library(scRNAseq)
out <- scRNAseq::listDatasets()
out %>% as.data.frame
```
```{r}
out %>% as.data.frame %>% dplyr::filter(str_detect(Part,'brain') & Taxonomy==10090)

out %>% as.data.frame %>% dplyr::filter(str_detect(Part,'stem') & Taxonomy==10090)
```
```{r message=FALSE, warning=FALSE}
marques2016oligodendrocyte <- MarquesBrainData(ensembl=TRUE)
```
```{r message=FALSE, warning=FALSE}
oligodendrocyte <- marques2016oligodendrocyte
keep_transcript <- rowSums(counts(oligodendrocyte) > 10) > 10
oligodendrocyte <- oligodendrocyte[keep_transcript, ]
oligodendrocyte %>% rowData() -> oligodendrocyte.transc
```
```{r}
buettner2015computational <- BuettnerESCData()
```
```{r}
ESC <- buettner2015computational
keep_transcript <- rowSums(counts(ESC) > 1) > 1
ESC <- ESC[keep_transcript, ]
ESC %>% rowData() -> ESC.transc
```


```{r}
library(org.Mm.eg.db)
org.Mm.eg.db %>% keytypes

ESC.ENTREZID             <- clusterProfiler::bitr(rownames(ESC.transc), "ENSEMBL", "ENTREZID", "org.Mm.eg.db")
oligodendrocyte.ENTREZID <- clusterProfiler::bitr(rownames(oligodendrocyte.transc), "ENSEMBL", "ENTREZID", "org.Mm.eg.db")

```


```{r}

ESC_ego <- clusterProfiler::enrichGO(gene         = ESC.ENTREZID$ENTREZID,
                OrgDb         = org.Mm.eg.db,
                keyType       = 'ENTREZID',
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.01,
                qvalueCutoff  = 0.01)
head(ESC_ego,10)    
```

```{r}
OLIGO_ego <- clusterProfiler::enrichGO(gene         = oligodendrocyte.ENTREZID$ENTREZID,
                OrgDb         = org.Mm.eg.db,
                keyType       = 'ENTREZID',
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.01,
                qvalueCutoff  = 0.01)
head(OLIGO_ego,10)  
```

```{r}
OLIGO_vs_ESC <- clusterProfiler::enrichGO(gene          = oligodendrocyte.ENTREZID$ENTREZID,
                universe      = ESC.ENTREZID$ENTREZID,
                OrgDb         = org.Mm.eg.db,
                keyType       = "ENTREZID",
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.01,
                qvalueCutoff  = 0.01,
        readable      = TRUE)

head(OLIGO_vs_ESC,10)
```

```{r}

oligo_mkk <- clusterProfiler::enrichMKEGG(gene = ESC.ENTREZID$ENTREZID,
                    organism = "mmu",
                   pvalueCutoff = 0.01,
                   qvalueCutoff = 0.01)

oligo_mkk@result
```

```{r}
mkk <- clusterProfiler::enrichMKEGG(gene = oligodendrocyte.ENTREZID$ENTREZID,
                   universe = ESC.ENTREZID$ENTREZID,
                    organism = "mmu",
                   pvalueCutoff = 0.01,
                   qvalueCutoff = 0.01)

mkk@result  
```

### ReactomePA

```{r}
#BiocManager::install("ReactomePA", checkBuilt = T)
ecs_reactome  <- ReactomePA::enrichPathway(ESC.ENTREZID$ENTREZID,
                          minGSSize = 15,  organism = "mouse", pvalueCutoff = 0.01)

ecs_reactome@result %>% dplyr::select(matches('descri'))
```


```{r}
oligo_ecs_reactome  <- ReactomePA::enrichPathway(oligodendrocyte.ENTREZID$ENTREZID,
                          universe = ESC.ENTREZID$ENTREZID,
                          minGSSize = 5,  organism = "mouse", pvalueCutoff = 0.01)
oligo_ecs_reactome@result%>% dplyr::select(matches('descri'))
```


### Disease over-representation analysis (human)

```{r}
out %>% as.data.frame() %>% dplyr::filter(str_detect(Part,'lung') & Taxonomy==9606)
out %>% as.data.frame() %>% dplyr::filter(str_detect(Part,'plur') & Taxonomy==9606)
```


```{r}
zilionis2019singlecell <- ZilionisLungData(ensembl = T)
lamanno2016molecular   <- LaMannoBrainData('human-ips', ensembl = T)
```

```{r}
lung      <- zilionis2019singlecell
human_ips <- lamanno2016molecular

keep_transcript <- rowSums(counts(lung) > 1) > 1
lung <- lung[keep_transcript, ]
lung %>% rowData() -> lung.transc


keep_transcript <- rowSums(counts(human_ips) > 1) > 1
human_ips <- human_ips[keep_transcript, ]
human_ips %>% rowData() -> human_ips.transc



lung.ENTREZID <- clusterProfiler::bitr(rownames(lung.transc), "ENSEMBL", "ENTREZID", "org.Hs.eg.db")

human_ips.ENTREZID <- clusterProfiler::bitr(rownames(human_ips.transc), "ENSEMBL", "ENTREZID", "org.Hs.eg.db")
```

```{r}

x <- DOSE::enrichDO(gene          = lung.ENTREZID$ENTREZID,
              ont           = "DO",
              pvalueCutoff  = 0.05,
              pAdjustMethod = "BH",
              minGSSize     = 5,
              maxGSSize     = 500,
              qvalueCutoff  = 0.05,
              readable      = FALSE)

x@result
```
```{r}

x <- DOSE::enrichDO(gene          = human_ips.ENTREZID$ENTREZID,
              ont           = "DO",
              pvalueCutoff  = 0.05,
              pAdjustMethod = "BH",
              minGSSize     = 5,
              maxGSSize     = 500,
              qvalueCutoff  = 0.05,
              readable      = FALSE)

x@result
```

### Over-representation analysis for the disease gene network


```{r}

dgn <-DOSE::enrichDGN(
  gene   = lung.ENTREZID$ENTREZID,
  pvalueCutoff = 0.01,
  pAdjustMethod = "BH",
  minGSSize = 400,
  maxGSSize = 700,
  qvalueCutoff = 0.01)


dgn %>% as.data.frame
```

### Universal enrichment analysis

```{r}
#devtools::install_dev("vroom")
cell_marker_data <- vroom::vroom('http://bio-bigdata.hrbmu.edu.cn/CellMarker/download/Human_cell_markers.txt')

## instead of `cellName`, users can use other features (e.g. `cancerType`)
cells <- cell_marker_data %>%
    dplyr::select(cellName, geneID) %>%
    dplyr::mutate(geneID = strsplit(geneID, ', ')) %>%
    tidyr::unnest()
```
```{r}
x <-  clusterProfiler::enricher(lung.ENTREZID$ENTREZID, TERM2GENE = cells)
x@result
```

#MSigDb analysis
Molecular Signatures Database is a collection of annotated gene sets. It contains 8 major collections:

    H: hallmark gene sets
    C1: positional gene sets
    C2: curated gene sets
    C3: motif gene sets
    C4: computational gene sets
    C5: GO gene sets
    C6: oncogenic signatures
    C7: immunologic signatures


```{r}
library(msigdbr)
msigdbr_show_species()
```

```{r}
m_df <- msigdbr(species = "Homo sapiens")


m_df  %>% as.data.frame %>% dplyr::select(gs_description) %>% unique %>% tail
```

```{r}
m_t2g <- msigdbr(species = "Homo sapiens", category = "C7") %>% 
  dplyr::select(gs_name, entrez_gene)
m_t2g
```

```{r}
em <- clusterProfiler::enricher(lung.ENTREZID$ENTREZID, TERM2GENE=m_t2g)
head(em)
```
### Biological theme comparison

Comparing multiple gene lists


```{r}

out %>% as.data.frame() %>% dplyr::filter(str_detect(Part,'embryonic stem cells') & Taxonomy==9606)
out %>% as.data.frame() %>% dplyr::filter(str_detect(Part,'induced pluripotent stem cells') & Taxonomy==9606)
out %>% as.data.frame() %>% dplyr::filter(str_detect(Part,'lung') & Taxonomy==9606)
out %>% as.data.frame() %>% dplyr::filter(str_detect(Part,'pancreas') & Taxonomy==9606)
out %>% as.data.frame() %>% dplyr::filter(str_detect(Part,'cortex') & Taxonomy==9606)
out %>% as.data.frame() %>% dplyr::filter(str_detect(Part,'.*') & Taxonomy==9606)
```


```{r}
Lung         <- ZilionisLungData(ensembl = T)
Human_ips        <- LaMannoBrainData('human-ips', ensembl = T)
Human_es         <-  LaMannoBrainData('human-es', ensembl = T)
Pancreas     <-       	MuraroPancreasData(ensembl = T)
#Cortex  <-  ReprocessedFluidigmData(ensembl=T)
PBMC    <- MairPBMCData(mode = 'rna', ensembl = T)
```
```{r}
keep_transcripts <- function(dataset, min_counts=10){keep_filter <- rowSums(counts(dataset) > min_counts) > min_counts
                                                   dataset[keep_filter, ] %>% rowData() -> filtered.transc
                                                   return(filtered.transc)}

my.bitr <- function(some_ens){some_ens %>% rownames %>% clusterProfiler::bitr( "ENSEMBL", "ENTREZID", "org.Hs.eg.db") %>% .[['ENTREZID']]}
purrr::compose(my.bitr,keep_transcripts) -> ENSEMBL_to_ENTREZID

list(Pancreas,PBMC,Human_es,Human_ips,Lung) %>% purrr::set_names(c('Pancreas','PBMC','Human_es','Human_ips','Lung')) -> my.cells

purrr::map(my.cells,ENSEMBL_to_ENTREZID) -> my_cells.ENTREZID


``` 

```{r fig.height=5, fig.width=15}
library(clusterProfiler)
library(ReactomePA)

ck <- clusterProfiler::compareCluster(geneCluster = my_cells.ENTREZID, fun = enrichPathway, pvalueCutoff=0.05)
dotplot(ck)
```


```{r fig.height=5, fig.width=15}
library(clusterProfiler)
library(ReactomePA)
ck2 <- compareCluster(geneCluster = my_cells.ENTREZID, fun = enrichGO, OrgDb = org.Hs.eg.db)
ck2 <- setReadable(ck2, OrgDb = org.Hs.eg.db, keyType="ENTREZID")
dotplot(ck2)
```

```{r fig.height=5, fig.width=15}
ck3 <- compareCluster(my_cells.ENTREZID, fun="enrichKEGG",
                     organism="hsa", pvalueCutoff=0.01)
dotplot(ck3)
```














