---
title: "Clase 01 de junio 2020: Recolección de datos RNA-seq, SQL y expresión diferencial."
output:
  html_document:
    df_print: paged
---

**Video**

Comienza en 11:03. Clase 01 junio 2020: Queries tipo SQL & expresión diferencial:

<iframe width="560" height="315" src="https://www.youtube.com/embed/gtJni7z9xTg" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>


**Descargar y descomprimir la base de datos**
```{r echo=TRUE, message=FALSE, warning=FALSE}
library(GEOquery)
library(SRAdb)
library(DBI)
setwd("~/clase_01_junio_2020")
#download.file('https://gbnci-abcc.ncifcrf.gov/backup/SRAmetadb.sqlite.gz',destfile = 'SRAmetadb.sqlite.gz')
#gunzip(filename= 'SRAmetadb.sqlite.gz', destname = 'SRAmetadb.sqlite', overwrite = T, remove = F)
```


```{r echo=TRUE, message=FALSE, warning=FALSE}
setwd("~/clase_01_junio_2020")
library(GEOquery)
library(SRAdb)
library(DBI)

sqlfile <- "/home/rstudio/clase_01_junio_2020/SRAmetadb.sqlite"  #catálogo!!!!!
sra_con <- dbConnect(SQLite(),sqlfile)
```


```{r echo=TRUE, message=FALSE, warning=FALSE}
#Interrogar al catálogo
sra_tables <- dbListTables(sra_con)
sra_tables
```
```{r echo=TRUE, message=FALSE, warning=FALSE}
dbListFields(sra_con,"submission")
```

**Desde una Run Accesion**

```{r echo=TRUE, message=FALSE, warning=FALSE}
#Una query (consulta) tipo sql
dbGetQuery(
  sra_con, paste( "SELECT  submission_alias, lab_name,broker_name,submission_comment,sradb_updated 
                   FROM    submission
                   WHERE","submission_accession 
                   LIKE   'SRA114259'",sep=" "))

```

```{r echo=TRUE, message=FALSE, warning=FALSE}
archivos.relacionados <- listSRAfile( c("SRA114259"), sra_con, fileType ='sra')
archivos.relacionados
```
**Obtener otros campos**

```{r echo=TRUE, message=FALSE, warning=FALSE}
conversion            <- sraConvert(c('SRA114259'), sra_con = sra_con )
archivos.relacionados <- apply(conversion, 2, unique)
archivos.relacionados
```

```{r echo=TRUE, message=FALSE, warning=FALSE}
estudios.relacionados <- archivos.relacionados$study
estudios.relacionados
```

```{r echo=TRUE, message=FALSE, warning=FALSE}
dbListFields(sra_con,"study")
dbListFields(sra_con,"run")

```


```{r echo=TRUE, message=FALSE, warning=FALSE}
mi.studio.1 <-dbGetQuery(
  sra_con,
paste("SELECT study_alias,study_accession,study_title,study_abstract,study_description,study_type",
      "FROM study 
       WHERE",      "study_accession 
       LIKE 'SRP191205'"
       ,sep=" "))

mi.studio.1      
```

**Obtención de estudios desde keywords **
buscar otras bases de datos: data-base query sql proteomics -> sqlite + R
```{r echo=TRUE, message=FALSE, warning=FALSE}
mis.studios <- 
dbGetQuery(
  sra_con,
paste("SELECT center_name,broker_name,study_alias,study_accession,study_title,study_abstract,study_description,study_type",
      "FROM study WHERE",      "((study_title LIKE  '%transcriptomic%'  AND
                                study_title  LIKE  '%characterization%' AND
                                study_title  LIKE  '%ipsc%Macrophages%')
                                OR
                                (study_title LIKE  '%monocyte-derived%'  AND
                                study_title  LIKE  '%TLR2/1 ligand%' AND
                                study_title  LIKE  '%MDMs%')
                                OR
                                (study_title LIKE '%Type I IFNs%'  AND
                                study_title LIKE  '%Macrophages%' AND
                                study_title LIKE  '%Epigenomic Landscape%'))
                                AND
                                (study_type LIKE 'transcr%')"
                                ,sep=" "))

mis.studios
```

```{r echo=TRUE, message=FALSE, warning=FALSE}
mis.studios$study_alias
```


```{r echo=TRUE, message=FALSE, warning=FALSE}
RNAseqs <- mis.studios$study_alias
RNAseqs
```

**Es particular para GEO**

##################
**Primer estudio**

`RNA-seq GSE100382`
##################

```{r echo=TRUE, message=FALSE, warning=FALSE}
#En geo esto es solo metadata: nombres de las muestas, protocolos, etc. no pero datos de expresión.
library(GEOquery)
library(dplyr)
library(Biobase)

selected.set = 'GSE100382'
i            = which(selected.set == RNAseqs)

downloaded.sets               = list()
downloaded.sets[[i]]          = getGEO(RNAseqs[i], GSEMatrix =TRUE, getGPL= T, AnnotGPL = T)
my.RNAseq                     = downloaded.sets[[i]] 
sampleNames                   = my.RNAseq[[1]]$title
sampleNamesGEO                = sampleNames(phenoData(my.RNAseq[[1]]))
data.frame(sampleNames,sampleNamesGEO)  %>% arrange(sampleNames) -> names.to.comvert
```

```{r echo=TRUE, message=FALSE, warning=FALSE}
#Bajamos los datos de expresión
library(stringr)
library(tibble)
#setwd("~/RNAseq_meta_analysis")
manydirectories = list.dirs()
directorynames  = basename(manydirectories)
if(selected.set %in% directorynames){unlink(selected.set, recursive = TRUE)}
files = getGEOSuppFiles('GSE100382')
rownames(files) %>% str_detect('.txt.gz') -> my.txt.gz
rownames(files)[my.txt.gz] %>% str_replace_all('.gz','') -> expr.file.path
gunzip(rownames(files)[my.txt.gz])
```

```{r echo=TRUE, message=FALSE, warning=FALSE}
#Igualar nombres de las muestras, que son los nombres de las columnas de la matriz de expresióm
seqdata = read.table(expr.file.path,header = T, sep = '\t')
first.col.genes = colnames(seqdata)[1]
seqdata         = seqdata[!duplicated(seqdata[first.col.genes]), ]
seqdata%>%remove_rownames()%>%column_to_rownames(first.col.genes)%>%data.matrix()%>% as.matrix() -> rna.seq.exprs.mat

rna.seq.exprs.mat[,order(colnames(rna.seq.exprs.mat))] -> rna.seq.exprs.mat

cols.main.data    = as.character(names.to.comvert$sampleNames)
cols.supplem.data = as.character(colnames(rna.seq.exprs.mat))

if (cols.main.data!=cols.supplem.data){print('Nombres de columnas diferentes!')} else {print('Cols OK')}
###########################################################################
names.to.comvert$sampleNames <- as.character(colnames(rna.seq.exprs.mat))
cols.main.data    = as.character(names.to.comvert$sampleNames)
cols.supplem.data = as.character(colnames(rna.seq.exprs.mat))

if (cols.main.data!=cols.supplem.data){print('Nombres de columnas diferentes!')} else {print('Cols OK')}
##########################################################################
colnames(rna.seq.exprs.mat) <-names.to.comvert$sampleNamesGEO
```

```{r echo=TRUE, message=FALSE, warning=FALSE}
#Integración de la metadata con los datos de expresión
ok.data                       = rna.seq.exprs.mat[, (rownames(as.data.frame(my.RNAseq[[1]])))]
datasets.ready                = list()
datasets.ready[[RNAseqs[i]]]  = ExpressionSet(assayData  = ok.data,
                                              phenoData  = phenoData(my.RNAseq[[1]]),
                                             featureData = annotatedDataFrameFrom(ok.data, byrow=TRUE),
                                              Annotation = annotation(my.RNAseq[[1]]),
                                          experimentData = experimentData(my.RNAseq[[1]]),
                                          protocolData   = protocolData(my.RNAseq[[1]]))
```

```{r echo=TRUE, message=FALSE, warning=FALSE}
datasets.ready %>% glimpse()
```
#####################
 **Segundo estudio**
 
`RNA-seq GSE55536`
#####################
```{r echo=TRUE, message=FALSE, warning=FALSE}
library(GEOquery)
library(dplyr)
library(Biobase)

selected.set = 'GSE55536'
i            = which(selected.set == RNAseqs)

downloaded.sets               = list()
downloaded.sets[[i]]          = getGEO(RNAseqs[i], GSEMatrix =TRUE, getGPL= T, AnnotGPL = T)
my.RNAseq                     = downloaded.sets[[i]] 
sampleNames                   = my.RNAseq[[1]]$title
sampleNamesGEO                = sampleNames(phenoData(my.RNAseq[[1]]))
data.frame(sampleNames,sampleNamesGEO)  %>% arrange(sampleNames) -> names.to.comvert
```

```{r echo=TRUE, message=FALSE, warning=FALSE}
library(stringr)
library(tibble)
#setwd("~/RNAseq_meta_analysis")
manydirectories = list.dirs()
directorynames  = basename(manydirectories)
if(selected.set %in% directorynames){unlink(selected.set, recursive = TRUE)}
files = getGEOSuppFiles(selected.set)
rownames(files) %>% str_detect('.txt.gz') -> my.txt.gz
rownames(files)[my.txt.gz] %>% str_replace_all('.gz','') -> expr.file.path
gunzip(rownames(files)[my.txt.gz])
```


```{r echo=TRUE, message=FALSE, warning=FALSE}
seqdata         = read.table(expr.file.path,header = T, sep = '\t')
first.col.genes = colnames(seqdata)[1]
seqdata         = seqdata[!duplicated(seqdata[first.col.genes]), ]
seqdata%>%remove_rownames()%>%column_to_rownames(first.col.genes)%>%data.matrix()%>% as.matrix() -> rna.seq.exprs.mat
rna.seq.exprs.mat[,order(colnames(rna.seq.exprs.mat))] -> rna.seq.exprs.mat

cols.main.data    = as.character(names.to.comvert$sampleNames)
cols.supplem.data = as.character(colnames(rna.seq.exprs.mat))
if (cols.main.data!=cols.supplem.data){print('Nombres de columnas diferentes!')} else {print('Cols OK')}
###########################################################################
colnames(rna.seq.exprs.mat) <- as.character(names.to.comvert$sampleNames)
###########################################################################
cols.main.data    = as.character(names.to.comvert$sampleNames)
cols.supplem.data = as.character(colnames(rna.seq.exprs.mat))
if (cols.main.data!=cols.supplem.data){print('Nombres de columnas diferentes!')} else {print('Cols OK')}
##########################################################################
colnames(rna.seq.exprs.mat) <-names.to.comvert$sampleNamesGEO
##########################################################################
```

```{r echo=TRUE, message=FALSE, warning=FALSE}
ok.data                       = rna.seq.exprs.mat[, (  rownames(as.data.frame(my.RNAseq[[1]]))  )]
###################################################################################################
#datasets.ready               = list()
###################################################################################################
datasets.ready[[RNAseqs[i]]]  = ExpressionSet(assayData  = ok.data,
                                              phenoData  = phenoData(my.RNAseq[[1]]),
                                             featureData = annotatedDataFrameFrom(ok.data, byrow=TRUE),
                                              Annotation = annotation(my.RNAseq[[1]]),
                                          experimentData = experimentData(my.RNAseq[[1]]),
                                          protocolData   = protocolData(my.RNAseq[[1]]))
```

```{r echo=TRUE, message=FALSE, warning=FALSE}
datasets.ready %>% glimpse()
```

##################
 **Tercer estudio**
 
`RNA-seq GSE82227`
##################

```{r echo=TRUE, message=FALSE, warning=FALSE}
library(GEOquery)
library(dplyr)
library(Biobase)

selected.set = 'GSE82227'
i            = which(selected.set == RNAseqs)

downloaded.sets               = list()
downloaded.sets[[i]]          = getGEO(RNAseqs[i], GSEMatrix =TRUE, getGPL= T, AnnotGPL = T)
my.RNAseq                     = downloaded.sets[[i]] 
sampleNames                   = my.RNAseq[[1]]$title
sampleNamesGEO                = sampleNames(phenoData(my.RNAseq[[1]]))
data.frame(sampleNames,sampleNamesGEO)  %>% arrange(sampleNames) -> names.to.comvert
```

```{r echo=TRUE, message=FALSE, warning=FALSE}
library(stringr)
library(tibble)
#etwd("~/RNAseq_meta_analysis")
manydirectories = list.dirs()
directorynames  = basename(manydirectories)
if(selected.set %in% directorynames){unlink(selected.set, recursive = TRUE)}
files = getGEOSuppFiles(selected.set)
rownames(files) %>% str_detect('.gz') -> my.txt.gz
rownames(files)[my.txt.gz] %>% str_replace_all('.gz','') -> expr.file.path
gunzip(rownames(files)[my.txt.gz])
```


```{r echo=TRUE, message=FALSE, warning=FALSE}
library(readr)
#############################################################################
#seqdata         = read.table(expr.file.path,header = T, sep = '\t')
################################################################################
seqdata          = read_csv(expr.file.path)
```


```{r echo=TRUE, message=FALSE, warning=FALSE}
first.col.genes = colnames(seqdata)[1]
seqdata         = seqdata[!duplicated(seqdata[first.col.genes]), ]
seqdata%>%remove_rownames()%>%column_to_rownames(first.col.genes)%>%data.matrix()%>% as.matrix() -> rna.seq.exprs.mat
rna.seq.exprs.mat[,order(colnames(rna.seq.exprs.mat))] -> rna.seq.exprs.mat

cols.main.data    = as.character(names.to.comvert$sampleNames)
cols.supplem.data = as.character(colnames(rna.seq.exprs.mat))
if (cols.main.data!=cols.supplem.data){print('Nombres de columnas diferentes!')} else {print('Cols OK')}
###########################################################################
colnames(rna.seq.exprs.mat) <- as.character(names.to.comvert$sampleNames)
###########################################################################
cols.main.data    = as.character(names.to.comvert$sampleNames)
cols.supplem.data = as.character(colnames(rna.seq.exprs.mat))
if (cols.main.data!=cols.supplem.data){print('Nombres de columnas diferentes!')} else {print('Cols OK')}
##########################################################################
colnames(rna.seq.exprs.mat) <-names.to.comvert$sampleNamesGEO
##########################################################################
```

```{r echo=TRUE, message=FALSE, warning=FALSE}
ok.data                       = rna.seq.exprs.mat[, (  rownames(as.data.frame(my.RNAseq[[1]]))  )]
###################################################################################################
#datasets.ready               = list()
###################################################################################################
datasets.ready[[RNAseqs[i]]]  = ExpressionSet(assayData  = ok.data,
                                              phenoData  = phenoData(my.RNAseq[[1]]),
                                             featureData = annotatedDataFrameFrom(ok.data, byrow=TRUE),
                                              Annotation = annotation(my.RNAseq[[1]]),
                                          experimentData = experimentData(my.RNAseq[[1]]),
                                          protocolData   = protocolData(my.RNAseq[[1]]))
```

```{r echo=TRUE, message=FALSE, warning=FALSE}
datasets.ready %>% glimpse()
```

Hasta ahora bajamos tablas de conteo (forma pre-procesada, ya con el mapeo listo! del RNA-seq). En nuestro caso los tres son illumina.

Si sabemos que las normalizaciones no son comparables hay que bajar el dato crudo.

Los 'reads' son datos crudos, que nosotros no hemos bajado ni bajamos a trabajar por requieren mucho tiempo y recursos.

##################
**Pre-procesamiento (mínimo y no recomendable, sólo fines proof-of-concept) y análisis de expresión diferencial**

`RNA-seq GSE100382`
##################

```{r echo=TRUE, message=FALSE, warning=FALSE}
#Terminar con números enteros positivos >=0
library(Biobase)
selected.ready.set <- datasets.ready$GSE100382
selected.ready.set %>% exprs() %>% round() -> countdata
head(countdata,2)
```
```{r echo=TRUE, message=FALSE, warning=FALSE}
library(stringr)
library(DESeq2)

coldata <- pData(selected.ready.set)
coldata$title 
```


```{r echo=TRUE, message=FALSE, warning=FALSE}
coldata$title %>% as.character() %>% str_replace_all('.*(_L_|_TL_).*','stimulated') %>% str_replace_all('.*(_N_|_T_).*','control') -> coldata$title
coldata$title
```


```{r echo=TRUE, message=FALSE, warning=FALSE}
ddsMat <- DESeqDataSetFromMatrix(countData = countdata,
                                 colData   = coldata,
                                 design    = ~ title)

keep <- rowSums(counts(ddsMat)) > 1
dds  <- ddsMat[keep,]
dds  <- estimateSizeFactors(dds)
```


```{r echo=TRUE, message=FALSE, warning=FALSE}
library(DESeq2)
DE.analysis  <- DESeq(dds)
res0         <- results(DE.analysis,   contrast = c('title','stimulated','control'))
Resul        <- lfcShrink(DE.analysis, contrast = c('title','stimulated','control'), res=res0)
```

```{r echo=TRUE, fig.height=7, fig.width=15, message=FALSE, warning=FALSE}
library(EnhancedVolcano)

EnhancedVolcano(Resul,
    lab = rownames(Resul),
    x = 'log2FoldChange',
    y = 'pvalue',
    xlim = c(-6,6),
    xlab = bquote(~Log[2]~ 'fold change'),
    pCutoff = 10e-10,
    FCcutoff = 2.0,
    pointSize = 2.0,
    labSize = 4.0,
    colAlpha = .8,
    legendPosition = 'right',
    legendLabSize = 12,
    legendIconSize = 4.0,
    drawConnectors = TRUE,
    widthConnectors = 0.2,
    colConnectors = 'grey30')
```

```{r echo=TRUE, message=FALSE, warning=FALSE}
mis.DEGs.studies                <- list()
mis.DEGs.studies[['GSE100382']] <- Resul
mis.DEGs.studies %>% .[['GSE100382']]   %>% glimpse()
```


##################
`RNA-seq GSE55536`
##################
```{r echo=TRUE, message=FALSE, warning=FALSE}
library(Biobase)
library(stringr)

selected.ready.set = datasets.ready$GSE55536
selected.ready.set %>% exprs() %>% round() -> countdata #Por qué redondear

coldata <- pData(selected.ready.set)
coldata$title 
```


```{r echo=TRUE, message=FALSE, warning=FALSE}
coldata$title %>% as.character() %>% str_replace_all('.*(M1|M2).*','stimulated') %>% str_replace_all('.*rep.*','control') -> coldata$title 
coldata$title
```

```{r echo=TRUE, message=FALSE, warning=FALSE}
library(DESeq2)
ddsMat <- DESeqDataSetFromMatrix(countData = countdata,
                                 colData   = coldata,
                                 design    = ~ title)
keep <- rowSums(counts(ddsMat)) > 1
dds  <- ddsMat[keep,]
dds  <- estimateSizeFactors(dds)
```

```{r echo=TRUE, message=FALSE, warning=FALSE}
library(DESeq2)

DE.analysis  <- DESeq(dds)
res0         <- results(DE.analysis,   contrast = c('title','stimulated','control'))
Resul        <- lfcShrink(DE.analysis, contrast = c('title','stimulated','control'), res=res0)
```

```{r echo=TRUE, fig.height=7, fig.width=15, message=FALSE, warning=FALSE}
library(EnhancedVolcano)

EnhancedVolcano(Resul,
    lab = rownames(Resul),
    x = 'log2FoldChange',
    y = 'pvalue',
    xlim = c(-6,6),
    xlab = bquote(~Log[2]~ 'fold change'),
    pCutoff = 10e-10,
    FCcutoff = 2.0,
    pointSize = 2.0,
    labSize = 4.0,
    colAlpha = .8,
    legendPosition = 'right',
    legendLabSize = 12,
    legendIconSize = 4.0,
    drawConnectors = TRUE,
    widthConnectors = 0.2,
    colConnectors = 'grey30')
```
```{r echo=TRUE, message=FALSE, warning=FALSE}
mis.DEGs.studies[['GSE55536']] <- Resul
mis.DEGs.studies %>% glimpse()
```


##################
`RNA-seq GSE82227`
##################
```{r echo=TRUE, message=FALSE, warning=FALSE}
library(Biobase)
library(stringr)

selected.ready.set = datasets.ready$GSE82227
selected.ready.set %>% exprs() %>% round() -> countdata #Por qué redondear

coldata <- pData(selected.ready.set)
coldata$characteristics_ch1.2 
```


```{r echo=TRUE, message=FALSE, warning=FALSE}
coldata$characteristics_ch1.2  %>%as.character()%>%str_replace_all('.*alone.*','control') %>% 
  str_replace_all('.*(ligand|gamma).*','stimulated') -> coldata$my.experiments

coldata$my.experiments
```
```{r echo=TRUE, message=FALSE, warning=FALSE}
library(DESeq2)
ddsMat <- DESeqDataSetFromMatrix(countData = countdata,
                                 colData   = coldata,
                                 design    = ~ my.experiments)

keep <- rowSums(counts(ddsMat)) > 1
dds  <- ddsMat[keep,]
dds  <- estimateSizeFactors(dds)
```

```{r echo=TRUE, message=FALSE, warning=FALSE}
library(DESeq2)

DE.analysis  <- DESeq(dds)
res0         <- results(DE.analysis,   contrast = c('my.experiments','stimulated','control'))
Resul        <- lfcShrink(DE.analysis, contrast = c('my.experiments','stimulated','control'), res=res0)
```

```{r echo=TRUE, fig.height=7, fig.width=15, message=FALSE, warning=FALSE}
library(EnhancedVolcano)

EnhancedVolcano(Resul,
    lab = rownames(Resul),
    x = 'log2FoldChange',
    y = 'pvalue',
    xlim = c(-6,6),
    xlab = bquote(~Log[2]~ 'fold change'),
    pCutoff = 10e-10,
    FCcutoff = 2.0,
    pointSize = 2.0,
    labSize = 2.0,
    colAlpha = .8,
    legendPosition = 'right',
    legendLabSize = 12,
    legendIconSize = 4.0,
    drawConnectors = TRUE,
    widthConnectors = 0.2,
    colConnectors = 'grey30')
```
```{r echo=TRUE, message=FALSE, warning=FALSE}
mis.DEGs.studies[['GSE82227']] <- Resul
mis.DEGs.studies %>% glimpse()
```



