---
{title: "Minado de conocimiento biologico"
date: 2020-11-12,
author: "Alejandro Acevedo-Aracena, _PhD_^[_FONDECYT Postdoctoral Fellow_, Universidad de Chile, deepen.data@gmail.com]; Manuel Muñoz-González^[_Pregrado_, Universidad de Chile]", output: {html_document: {theme: flatly, highlight: tango, toc: true, toc_float: {collapsed: false, smooth_scroll: true}, df_print: paged, code_folding: show}, html_notebook: default}}
---

<details>
<summary>**Instalaciones previas**</summary>

Corran este Chunk si no tienen estos paquetes ya instalados. 

```{r`Intalaciones iniciales, eval=F}

bioPacks <- c(# PAQUETES DESDE BIOCONDUCTOR
"Biostrings", # Manipulación de strings en biología
"biomaRt")    # Acceso a bases de datos biomedicas

instalables.bioPacks <- bioPacks[!(bioPacks %in% installed.packages()[,"Package"])] # Lista
BiocManager::install(instalables.bioPacks, ask=FALSE)
```

</summary>

```{r eval=F}
BiocManager::install("AnnotationHub", checkBuilt = T)
BiocManager::install(c("GOSemSim", "clusterProfiler", "DOSE", "ape", "ggtree", "org.Hs.eg.db", "org.Mn.eg.db"))
BiocManager::install("STRINGdb", checkBuilt = T)
install.packages("igraph", dependencies = T)
BiocManager::install("DESeq2", checkBuilt = T)
BiocManager::install("apeglm", checkBuilt = T)
BiocManager::install("pathview", checkBuilt = T)
```


### Data

```{r message=FALSE, warning=FALSE}
library(tidyverse)
data0 <- GEOquery::getGEO("GSE148349", GSEMatrix = T, AnnotGPL = T, getGPL = T)
data0 %>%
  names() %>%
  str_extract("GSE\\d+.GPL\\d+") %>%
  set_names(data0, .) -> data
```

GEOfile vs GEO

```{r message=FALSE, warning=FALSE}
data0 %>%
  names() %>%
  str_extract("GSE\\d+.GPL\\d+") %>%
  set_names(data, .) -> data # TODO: what?

# Posiblemente remover la primera cosa, porque esto es un best practice para evitar descargas
my.GEOfile <- GEOquery::getGEOfile("GSE148349", AnnotGPL = T, destdir = getwd())

# Existe localmente el archivo?
if (!"GSE148349" %in% list.files()) {
  suppfiles <- GEOquery::getGEOSuppFiles("GSE148349", makeDirectory = TRUE, baseDir = getwd())
  suppfiles %>% row.names() -> full.path.raw_tar
} else {
  full.path.raw_tar <- file.path(getwd(), "GSE148349", "GSE148349_RAW.tar")
}
```


```{r Unizpeado, message=FALSE, warning=FALSE}
# Ruta completa a la carpeta con la descarga
full.path.raw_tar %>% str_match(".*(?=GSE\\d+.RAW.*)") -> gse.folder

# Descomprimiendo el archivo dentro de esa carpeta
untar(tarfile = full.path.raw_tar, exdir = gse.folder)
# Eliminar archivo tar original?
```

```{r message=FALSE, warning=FALSE}
na.cleaner <- purrr::compose(na.omit, as.character)
my.gunzip <- function(a.file) {
  GEOquery::gunzip(a.file, remove = F, overwrite = T)
}
get.my.gz.files <- function(a.gse.folder) {
  a.gse.folder %>%
    list.files(full.names = T) %>%
    str_extract(".*gz") %>%
    na.cleaner() %>%
    map(my.gunzip)
}

# Luego deberiamos guardar esto como objetos RDS, por la compresión más eficiente
```

# Conviertiendo TSV a data.frame

```{r message=FALSE, warning=FALSE, include=FALSE, echo=TRUE}
gse.folder %>% get.my.gz.files()
gse.folder %>%
  list.files(full.names = T) %>%
  str_extract(".*tsv$") %>%
  na.cleaner() -> tsv.files.path


tsv.files.path %>%
  map(function(files.path00) {
    readr::read_delim(files.path00, delim = "\t", skip = 0, col_names = TRUE)
  }) -> df_list # Genera una lista de datafrmaes que corresponde a la salida de los TSV
# Estos dataframes deberian ser guardados para posteridad
# Lo demas puede morir
```

En este momento tenemos 66 archivos, con 66,000 filas en total.

```{r message=FALSE, warning=FALSE}
# Recuperando la infromación de GSM<->TSV
tsv.files.path %>%
  str_extract("GSM\\d+") %>%
  set_names(df_list, .) -> df_GSM_list

# Una función iterativa para extraer los GSM
get_gsm <- function(my.list.item) {
  gsm_code <- names(my.list.item)
  my.list.item %>%
    .[[1]] %>%
    dplyr::select(target_id, est_counts) %>%
    set_names(c("target_id", gsm_code)) -> a_gsm
  return(a_gsm)
}

# Crando una lista para guardar los GSM
list_of_gsms <- list() # (está vacia)

# Aplicamos iterativamente la función en un for
# TODO: deberiamos usar programación funcional
for (i in 1:length(df_GSM_list))
{
  df_GSM_list[i] %>% get_gsm() -> one.gsm
  list_of_gsms[[i]] <- one.gsm
}

# Reducimos la lista, haciendo un join para tener GSM unicos
# Construye una matris de expresión de todos los suplementarios
list_of_gsms %>% reduce(full_join) -> all_gsms
```


```{r message=FALSE, warning=FALSE}
# Exploración manual de los objetos
data$`GSE148349-GPL24247`@phenoData@data # Metadatos de las muestras
data$`GSE148349-GPL24247`@featureData@data # Metadatos de los genes
```


```{r message=FALSE, warning=FALSE}
library(Biobase)
data$`GSE148349-GPL24247` %>%
  exprs() %>%
  colnames() -> GSE148349_GPL24247 #
data$`GSE148349-GPL24676` %>%
  exprs() %>%
  colnames() -> GSE148349_GPL24676
```


```{r message=FALSE, warning=FALSE}
library(magrittr)
all_gsms %>%
  select(GSE148349_GPL24247) %>%
  as.matrix() %>%
  set_rownames(all_gsms$target_id) %>%
  as.data.frame() %>%
  drop_na() -> exprs_GSE148349_GPL24247
all_gsms %>%
  select(GSE148349_GPL24676) %>%
  as.matrix() %>%
  set_rownames(all_gsms$target_id) %>%
  as.data.frame() %>%
  drop_na() -> exprs_GSE148349_GPL24676
```

```{r}
# Construcción de un objeto expression set con la data fenotipica, anotación, experimentos, protocolos
exp_set_GSE148349_GPL24247 <- ExpressionSet(
  assayData = as.matrix(exprs_GSE148349_GPL24247),
  phenoData = data$`GSE148349-GPL24247`@phenoData,
  Annotation = data$`GSE148349-GPL24247`@annotation,
  experimentData = data$`GSE148349-GPL24247`@experimentData,
  protocolData = data$`GSE148349-GPL24247`@protocolData
)
```


```{r}
# Repitiendo para la otra cosa
exp_set_GSE148349_GPL24676 <- ExpressionSet(
  assayData = as.matrix(exprs_GSE148349_GPL24676),
  phenoData = data$`GSE148349-GPL24676`@phenoData,
  Annotation = data$`GSE148349-GPL24676`@annotation,
  experimentData = data$`GSE148349-GPL24676`@experimentData,
  protocolData = data$`GSE148349-GPL24676`@protocolData
)
```

```{r}
# Objeto construido, podemos uar Biobase expres
Biobase::exprs(exp_set_GSE148349_GPL24247)
# TODO: que la salida sea menos fea
```


```{r}
exp_set_GSE148349_GPL24247 %>%
  exprs() %>%
  round() -> countdata # Por qué redondear

coldata <- pData(exp_set_GSE148349_GPL24247)

coldata$title %>% str_detect("Untreated") -> untreated_bool

coldata$title[untreated_bool] <- "control"
coldata$title[!untreated_bool] <- "treatment"
```

--deberiamos usar lima porque DESeq2 es pesadisimo--

```{r message=FALSE, warning=FALSE}
# BiocManager::install("DESeq2", checkBuilt = T)

library(DESeq2)
ddsMat <- DESeqDataSetFromMatrix(
  countData = countdata,
  colData = coldata,
  design = ~title
)
keep <- rowSums(counts(ddsMat)) > 1
dds <- ddsMat[keep, ]
dds <- estimateSizeFactors(dds)
```

```{r}
# BiocManager::install("apeglm", checkBuilt = T)

# Analisis de expresión diferencial
DE.analysis <- DESeq(dds)
res0 <- results(DE.analysis, contrast = c("title", "treatment", "control"))
```
```{r eval=FALSE, include=FALSE}
# TODO: instalar VolcanoPlot
library(EnhancedVolcano)

EnhancedVolcano(res0,
  lab = rownames(res0),
  x = "log2FoldChange",
  y = "padj",
  xlim = c(-6, 6),
  xlab = bquote(~ Log[2] ~ "fold change"),
  pCutoff = 1e-5,
  FCcutoff = 2.0,
  pointSize = 2.0,
  labSize = 4.0,
  colAlpha = .8,
  legendPosition = "right",
  legendLabSize = 12,
  legendIconSize = 4.0,
  drawConnectors = TRUE,
  widthConnectors = 0.2,
  colConnectors = "grey30"
)
```


### Mining the biological knowledge of the differentially expressed genes

```{r}
# Seleccionamos solo los bojetos con un padj menor a 1e-7, y un fold chanche mayor a 4
# TODO: que es padj?
res0 %>%
  as.data.frame() %>%
  rownames_to_column("genes") %>%
  select(genes, padj, log2FoldChange) %>%
  filter(padj < 1.0e-07 & abs(log2FoldChange) > 4) -> my_DEGs
my_DEGs
```

### Obtener especie,  Symbol y Entrez ID

```{r}
library(org.Mm.eg.db) # TODO: no esta instalada en R 4+
library(org.Hs.eg.db) # Posible solución cambiando por Hs
org.Mm.eg.db %>% keytypes()

# Especie --falla por versión de R y Mm.eg.db
exp_set_GSE148349_GPL24247 %>%
  pData() %>%
  dplyr::select(starts_with("orga")) %>%
  .[[1]] %>%
  unique()

genes.ENTREZID <- clusterProfiler::bitr(my_DEGs$genes, "ENSEMBLTRANS", "ENTREZID", "org.Mm.eg.db")
DEGs0 <- inner_join(my_DEGs, genes.ENTREZID, by = c("genes" = "ENSEMBLTRANS"))
genes.SYMBOL <- clusterProfiler::bitr(my_DEGs$genes, "ENSEMBLTRANS", "SYMBOL", "org.Mm.eg.db")
DEGs <- inner_join(DEGs0, genes.SYMBOL, by = c("genes" = "ENSEMBLTRANS"))
DEGs
```
```{r}
library(GOSemSim)

MnGO <- godata("org.Mm.eg.db", ont = "BP")
# eg  <- DEGs$SYMBOL
eg <- DEGs$ENTREZID
sim <- mgeneSim(eg, semData = MnGO, measure = "Wang", combine = "BMA", verbose = FALSE)
DOSE::simplot(sim)
```

```{r}
sim %>% row.names() -> my.entrez.ids
DEGs %>% filter(ENTREZID == my.entrez.ids) -> my.dict # %>% .[['SYMBOL']] -> my.symbols
sim[my.dict$ENTREZID, my.dict$ENTREZID] -> sim2
row.names(sim2) <- my.dict$SYMBOL
colnames(sim2) <- my.dict$SYMBOL
DOSE::simplot(sim2)
```

### Gene cluster semantic similarity measurement

```{r}
# TODO: calcular distancia semantica para plots
# TODO: Más analsisi de distancia semantica entre dos genes, para ver como se comportan grupos de genes

# Usando filtros
DEGs %>%
  filter(log2FoldChange < 15) %>%
  .[["ENTREZID"]] -> gs1
DEGs %>%
  filter(log2FoldChange > 20) %>%
  .[["ENTREZID"]] -> gs2

# Distancia semantica entre gurpos de genes
clusterSim(gs1, gs2, semData = MnGO, measure = "Wang", combine = "BMA")
```

### GO enrichment analysis
```{r}
library(clusterProfiler)
library(org.Mm.eg.db)

# TODO: Ale comenta esto!
ggo <- groupGO(
  gene = DEGs$ENTREZID,
  OrgDb = org.Mm.eg.db,
  keyType = "ENTREZID",
  ont = "CC",
  level = 3,
  readable = TRUE
)

# Componentes celulares que se ven alterados
ggo %>% as.data.frame()
```
### GO enrichment analysis

```{r}
library(clusterProfiler)
data(geneList, package = "DOSE")
gene <- names(geneList)[abs(geneList) > 2]
# Entrez gene ID
head(gene)

# Muy pocos genes, deberian ser 22 vs 500
```


```{r}
library(org.Hs.eg.db)
glimpse(gene)
glimpse(names(geneList))
ego <- enrichGO(
  gene = gene,
  universe = names(geneList),
  OrgDb = org.Hs.eg.db,
  keyType = "ENTREZID",
  ont = "CC",
  pAdjustMethod = "BH",
  pvalueCutoff = 0.01,
  qvalueCutoff = 0.05,
  readable = TRUE
)
head(ego)

# TODO: porque aparece/no aparece universe = ''?
```
```{r}
gene.df <- bitr(gene,
  fromType = "ENTREZID",
  toType = c("ENSEMBL", "SYMBOL"),
  OrgDb = org.Hs.eg.db
)

ego2 <- enrichGO(
  gene = gene.df$ENSEMBL,
  OrgDb = org.Hs.eg.db,
  keyType = "ENSEMBL",
  ont = "CC",
  pAdjustMethod = "BH",
  pvalueCutoff = 0.01,
  qvalueCutoff = 0.05
)
head(ego2, 3)
```

GO Gene Set Enrichment Analysis
```{r}
ego3 <- gseGO(
  geneList = geneList,
  OrgDb = org.Hs.eg.db,
  ont = "CC",
  minGSSize = 100, # Minimo 100 genes
  maxGSSize = 500, # Maximo 500 genes
  pvalueCutoff = 0.05,
  verbose = FALSE
)

ego3 %>% head()
geneList %>% length()
ego3@result %>%
  select(setSize) %>%
  colSums()

# TODO: Describir como el universo cambia segun hipotesis biologica
```


```{r fig.height=4, fig.width=8}
goplot(ego)
```

```{r fig.height=4, fig.width=8}
goplot(ego2)
```

# KEGG

```{r}
# Analisis de enrequecimiento contra KEGG
data(geneList, package = "DOSE")
gene <- names(geneList)[abs(geneList) > 2]

kk <- enrichKEGG(
  gene = gene,
  organism = "hsa",
  pvalueCutoff = 0.05
)

kk@result
```

```{r}
kk2 <- gseKEGG(
  geneList = geneList,
  organism = "hsa",
  minGSSize = 120,
  pvalueCutoff = 0.05,
  verbose = FALSE
)
kk2@result
```


## KEGG module over-representation analysis
KEGG Module is a collection of manually defined function units. In some situation, KEGG Modules have a more straightforward interpretation.

```{r}
mkk <- enrichMKEGG(
  gene = gene,
  organism = "hsa",
  pvalueCutoff = 1,
  qvalueCutoff = 1
)
mkk@result
```

```{r}
mkk2 <- gseMKEGG(
  geneList = geneList,
  organism = "hsa",
  pvalueCutoff = 1
)
mkk2@result
```


### Visualize enriched KEGG pathways

```{r}
# Esto abre una ventana online
clusterProfiler::browseKEGG(kk, "hsa04110")
```

```{r}
# BiocManager::install("pathview", checkBuilt = T)

library("pathview")
hsa04110 <- pathview(
  gene.data = geneList,
  pathway.id = "hsa04110",
  species = "hsa",
  limit = list(gene = max(abs(geneList)), cpd = 1)
)
```

## GETTING INTERACTIONS WITH STRING 

[STRING](https://www.string-db.org) is a database of known and predicted protein-protein interactions

```{r eval=F}
library(STRINGdb)
data(interactions_example)
data(diff_exp_example1)

# To begin, you should first know the NCBI taxonomy identifiers of the organism on which you have performed the experiment (e.g. 9606 for Human, 10090 for mouse).
# create a new STRING_db object
string_db <- STRINGdb$new(version = "11", species = 9606, score_threshold = 0, input_directory = "")

tp53 <- string_db$mp("tp53")
atm <- string_db$mp("atm")

string_db$get_neighbors(c(tp53, atm))

string_db$get_neighbors(tp53)
```
