---
title: "R Notebook"
output: html_notebook
---

# 

[RPub de esta clase](https://rpubs.com/DeepenData/665531)

## Manipulando data local

```{r Chunk que carga packs de datos}
#install.packages("devtools", dependencies = TRUE)
#install.packages("curl", dependencies = TRUE)
#install.packages("tidyverse", dependencies = TRUE)
library(readr)
my.data <- mtcars

# Exportando data
write_csv(my.data,  path = 'mtcars.csv' ) # Coma Separated Value
write_tsv(my.data,  path = 'mtcars.tsv')  # Tab Separated Value
write_delim(my.data,path = 'mtcars.txt', delim = ";") # Texto con delimitador ";"
write_delim(my.data,path = 'mtcars.hola', delim = "*_*") # Texto con delimitador "*_*"

openxlsx::write.xlsx(my.data, file = "mtcars.xlsx") # Excel
```

TODO: Cual es la utilidad de separadores arbitrarios como \*_\* ?

```{r Importando Data local, message=FALSE}
# (Re)Importando data
mtcars.csv<-read_csv('mtcars.csv')
mtcars.tsv<-read_tsv('mtcars.tsv')
mtcars.txt<-read_delim('mtcars.txt', delim = ";")
mtcars.hola<-read_delim('mtcars.hola', delim = "*_*")

library(readxl) # Importando un Excel
mtcars.xlsx<-read_excel('mtcars.xlsx')
```

## Manipulando columna rownames

```{r}
library(tibble)   # TODO: que hace?
rownames_to_column(my.data, var = "Car brand")   -> my.data.2
library(magrittr) # TODO: que hace?
my.data %>% rownames_to_column(var = "Car brand")-> my.data.2

#####Con pipes#############
my.data.2 %>%   write_csv('mtcars.csv')
my.data.2 %>%   write_tsv('mtcars.tsv')
my.data.2 %>% write_delim('mtcars.txt',  delim = ";")
my.data.2 %>% write_delim('mtcars.hola', delim = "*_*")
my.data.2 %>% openxlsx::write.xlsx("mtcars.xlsx")

###########Lectura##################
read_csv('mtcars.csv', skip = 0)                         ->mtcars.csv
read_tsv('mtcars.tsv', skip = 0)                         ->mtcars.tsv
read_delim('mtcars.txt', delim = ";", skip = 0)          ->mtcars.txt
read_delim('mtcars.hola', delim = "*_*", skip = 0)       ->mtcars.hola
read_excel('mtcars.xlsx', skip = 0)                      ->mtcars.xlsx
str(mtcars.xlsx)

########Chequear igualdad###############
library(dplyr)
all_equal(mtcars.csv,mtcars.tsv)   # La data del csv deberia ser igual a la del tsv
all_equal(mtcars.txt,mtcars.hola)  # La data del txt deberia ser igual a la del .hola
all_equal(mtcars.hola,mtcars.xlsx) # La data del txt deberia ser igual a la del Excel

list(mtcars.csv, mtcars.tsv, mtcars.txt, mtcars.hola,mtcars.xlsx) %>% unique() %>% length() # TODO: porque da dos salidas? Una es una tabla y el otro es una consola?

```

## Descargando data online

```{r Descargando data en linea}
library(utils)
#getwd()

###GitHub
my.github.url <- 'https://github.com/DeepenData/Computational-Biology-and-Bioinformatics/raw/master/Wang2018_supplemental_DEGs.xlsx'

download.file(url= my.github.url,  destfile= 'Wang2018_supplemental_DEGs.xlsx')
#hola.github <- read_excel("Wang2018_supplemental_DEGs.xlsx", skip = 1) # ERROR!

###Dropbox
my.dropbox.url <- 'https://www.dropbox.com/s/j3kiivpcbghpb4v/log2FC.csv'

#download.file(url= my.dropbox.url,  destfile= 'log2FC.csv', method = "wget") # ERROR!
hola.dropbox <- read_csv("log2FC.csv")

###Sitio web arbitrario
my.supplementary.url <- 'https://www.pnas.org/highwire/filestream/794560/field_highwire_adjunct_files/0/pnas.1800165115.sd01.xlsx'

#download.file(url= my.supplementary.url,  destfile= 'pnas.1800165115.sd01.xlsx', method = "wget") # ERROR!
hola.supplementary <- read_excel("pnas.1800165115.sd01.xlsx")
```

```{r Leyendo directamente data en linea, paged.print=TRUE}
read_csv('https://raw.githubusercontent.com/DeepenData/Computational-Biology-and-Bioinformatics/master/labels.csv')

# Esto es particularmente util si usan Google Sheets
# TODO: hacer una GoogleSheet.csv para demostraciones
```
## Datos omicos

```{r Instalación de pre-requisitos Bioconductor}
install.packages("BiocManager", dependencies = TRUE)

BiocManager::install()
BiocManager::install("Biostrings")
BiocManager::install("biomaRt")
install.packages("biomartr")#, dependencies = TRUE)
```

```{r Datos Omicos, warning=FALSE, paged.print=TRUE}
library(biomartr)
library(magrittr)
#Objetos clase DNAStringSet
inmortal <- getGenome( db       = "refseq", #Más bases de datos
                       organism = "Thermococcus gammatolerans",
                       path     = file.path("_ncbi_downloads","genomes") ) %>%  read_genome()
inmortal
```
<!-- WLS2 es malo por EEE (Embrace, Extend and Extinguish o EEE) -->
```{r Datos Omicos Humanos, warning=FALSE, paged.print=TRUE, eval=FALSE}
human <- read_genome("/home/alejandro/DeepenData/Taller_sept_oct_nov_dic_2020/_ncbi_downloads/genomes/Homo_sapiens_genomic_refseq.fna.gz") # TODO: Corregir dirección local
human 

human %>% names() %>% sample(10)
```

## Epigenomica

```{r Instala GEOquery, message=FALSE, warning=FALSE}
#if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")

#BiocManager::install("GEOquery")
library(GEOquery)
```

```{r Descargando datasets GEO, paged.print=TRUE}
gset = getGEO("GSE36278") # Descarga un dataset de
fData(gset[[1]])          # Lista las sondas del set
```

## Transcriptomica

```{r Descarga datasets de transcriptomica, paged.print=TRUE}
inmortal <- getRNA( db       = "refseq", #Más bases de datos
                       organism = "Thermococcus gammatolerans",
                       path     = file.path("_ncbi_downloads","transcriptome") ) %>%  read_rna()

inmortal %>% names() %>% sample(10)
```

## Proteomica

```{r Descarga datasets de proteomica}
download_species <- c("Thermococcus gammatolerans", 
                      "Thermotoga maritima")
# retrieve these three species from NCBI RefSeq                       
biomartr::getProteomeSet("refseq", organisms = download_species, path = "set_proteomes")

read_proteome("/home/alejandro/DeepenData/Taller_sept_oct_nov_dic_2020/set_proteomes/Tmaritima.faa") # TODO: deslocalizar
```

Proteómica Importing de [ProteomExchange](http://www.proteomexchange.org/) [R for Proteomics]( https://bioconductor.org/packages/release/data/experiment/vignettes/RforProteomics/inst/doc/RforProteomics.html)

## Estudios diferenciales

```{r Instalando RforProteomics, warning=FALSE}
BiocManager::install()
BiocManager::install("RforProteomics",
                     ask = F, 
                     dependencies = TRUE,
                     type = "source",
                     checkBuilt = TRUE)
```

```{r Corriendo RforProteomics}
## Experiment information
library("rpx")
px1 <- PXDataset("PXD000001")
pxfiles(px1)
```

follow: 

```
sudo apt-cache search libnetcdf 
sudo apt-get update 
sudo apt-get install libnetcdf-c++4-1
```

[Cartografos](https://rtask.thinkr.fr/installation-of-r-4-0-on-ubuntu-20-04-lts-and-tips-for-spatial-packages/)

```{r}
#BiocManager::install("MSnbase", ask = T, dependencies = TRUE,   type = "source", checkBuilt = TRUE)
library(magrittr)
library(Biobase)
library(MSnbase)

## Downloading the mzTab data
mztab <- pxget(px1, "PXD000001_mztab.txt")
qnt <- readMzTabData(mztab, what = "PEP", version = "0.9")
```

<!-- Karina tiene problemas con XCode porque Apple es Apple -->
