---
title: "Programación funcional"
date: 2020-10-29
author: 'Acevedo, A. & Munoz, M. U. de Chile.'
output:
  html_document:
    df_print: paged
---

```{r message=FALSE, warning=FALSE}
library(tidyverse)
data <- GEOquery::getGEO('GSE148349',GSEMatrix = TRUE, AnnotGPL = T, getGPL = TRUE )
class(data)
data
```

```{r message=FALSE, warning=FALSE}
data %>% names %>% str_replace('_.*gz','') -> new.names 


library(magrittr)

data %<>% purrr::set_names(new.names)
```

```{r message=FALSE, warning=FALSE}
library(Biobase)

data %>% names

lapply(data, slotNames)


purrr::map(data, slotNames)

#Usar una función de Biobase para extraer info
purrr::map(data, pData)

library(purrr)

map(data, experimentData) %>% map(slotNames)

not_a_na          <- purrr::compose(na.omit , as.character)

A <- c(NA,'ddd','dgrfg',NA)
A %>% na.omit %>% as.character
A %>% not_a_na
```


```{r message=FALSE, warning=FALSE}
library(stringr)


list.of.phenoData <- map(data, pData)

list.of.phenoData %>% map(names) %>% map(~str_extract(.,'org.*'))  %>% map(not_a_na) -> my.organism.entry

map(list.of.phenoData , purrr::pluck('organism_ch1'))
#map(list.of.phenoData , pluck(my.organism.entry[[1]]))
```

```{r}
#exprs.matrices.list <- tibble::lst(`Mus musculus` = NULL, `Homo sapiens` = NULL)

```

### Desde los ID de las instancias hasta los mecanismos (desde ensembl_transcript_ID hasta EC number, para reconstruir redes)

```{r message=FALSE, warning=FALSE}

my.GEOfile = GEOquery::getGEOfile('GSE148349',AnnotGPL = T, destdir = getwd())

GEOquery::getGEO(filename=my.GEOfile) -> soft


suppfiles = GEOquery::getGEOSuppFiles('GSE148349')
suppfiles %>% row.names() -> full.path.raw_tar
full.path.raw_tar %>% str_match('GSE\\d+.RAW.*') 
full.path.raw_tar %>% str_match('.*(?=GSE\\d+.RAW.*)') -> gse.folder
untar(tarfile = full.path.raw_tar ,exdir =  gse.folder)
```
```{r message=FALSE, warning=FALSE}
na.cleaner      <- purrr::compose(na.omit,as.character)

my.gunzip       <- function(a.file){GEOquery::gunzip(a.file,remove = F, overwrite = T)}

get.my.gz.files <- function(a.gse.folder){a.gse.folder %>% list.files(full.names = T ) %>% 
                                          str_extract('.*gz') %>% na.cleaner %>% map(my.gunzip)
                                          }

gse.folder %>% get.my.gz.files

gse.folder %>% list.files(full.names = T ) %>% str_extract('.*tsv$') %>%  na.cleaner -> tsv.files.path

#Función anónima
tsv.files.path %>%
      map(function(files.path00){readr::read_delim(files.path00, delim = '\t', skip = 0,col_names = TRUE)}) -> df_list #

df_list %>% class
df_list %>% length
df_list %>% names
```


```{r message=FALSE, warning=FALSE}
tsv.files.path %>% str_extract('GSM\\d+') -> gsm.names

df_list %<>% purrr::set_names(gsm.names)

df_list %>% names
df_list['GSM4462281']

```

### GSMs sólo de Homo sapiens
```{r}

my.pData.list <- map(data, pData)

GSMs.by.gpl <- map( my.pData.list, pluck('geo_accession'))


GSMs.by.gpl[[2]] -> Homo.sapiens.samples
df_list[Homo.sapiens.samples] -> Homo.sapiens.samples.list

Homo.sapiens.samples.list %>% length()

Homo.sapiens.samples.list[[1]]
```

```{r}

map( Homo.sapiens.samples.list, pluck('target_id')) -> all.genes

all.genes %>% purrr::reduce(union) -> all.commom.genes

all.commom.genes %>% length

all.commom.genes %>% sample(5)
```


### 
```{r message=FALSE, warning=FALSE, eval=F}
BiocManager::install("KEGGREST")
library(KEGGREST)
```


```{r message=FALSE, warning=FALSE}
library(KEGGREST)
listDatabases()

#my.EC <- c('1.8.1.4')#, '2.7.4.12', '', '2.7.4.13', '2.7.4.9', '2.7.4.9') 
#keggFind("genes", my.EC) %>% as.matrix() %>% rownames()  %>%keggGet() -> found.kegg
#keggFind("genes",'1738')
#keggFind("genome",'ENSG00000091140')
```
### Convertir desde ensembl_transcript_id a entrezgene_id (biomaRt)
```{r message=FALSE, warning=FALSE}
library(KEGGREST)
library(biomaRt)
ensembl    <- useMart("ensembl") #para toda la base de datos
Hs.ensembl <- useMart("ensembl",dataset="hsapiens_gene_ensembl") #solo Humano

searchAttributes(mart = Hs.ensembl, pattern = 'GO')
```

```{r message=FALSE, warning=FALSE}
my.genes      <-   c('ENSG00000091140') #GENE como ejemplo

my.genes.symbols <- getBM(
attributes=c('ensembl_gene_id','hgnc_symbol','entrezgene_id','goslim_goa_description'),filters ='ensembl_gene_id',values = my.genes,mart = Hs.ensembl)

my.genes.symbols

```

```{r}


my.genes      <- c('ENST00000487058',all.commom.genes %>% sample(8), 'ENST00000535366')

my.genes.symbols <- getBM(
attributes=c('ensembl_transcript_id','hgnc_symbol','entrezgene_id','wikigene_description','goslim_goa_description'),filters ='ensembl_transcript_id',values = my.genes,mart = Hs.ensembl)

my.genes.symbols
```
```{r}
na.cleaner      <- purrr::compose(as.character, na.exclude)

my.genes.symbols$entrezgene_id %>% unique %>% na.cleaner %>% paste0('hsa:',.) %>% keggGet -> my.kegg.list

my.kegg.list %>% length
my.kegg.list %>% names

my.genes.symbols$hgnc_symbol[!is.na(my.genes.symbols$entrezgene_id)] %>% unique -> mis.nombres

my.kegg.list %>% purrr::set_names(mis.nombres)   -> my.kegg.info

my.kegg.info %>% names
my.kegg.info[["ENO2"]] %>% class

map( my.kegg.info, pluck('PATHWAY')) %>% glimpse()
map( my.kegg.info, pluck('PATHWAY')) %>% purrr::compact() 
```


```{r}


map( my.kegg.info, pluck('ORTHOLOGY')) %>% purrr::compact() %>% as.matrix() %>% as.data.frame() %>% unnest -> A


A %>%   mutate(`EC number` = ifelse(
                                    str_detect(V1, regex('\\[EC.\\d+' , ignore_case = T)), 
                                    'EC exists', 
                                      NA)) #%>% drop_na
```




